### PR TITLE
New checks in dockerlinter

### PR DESCRIPTION
**Pull request by phracek**
Fix PR contains help.md sanity checks based on CBP

### CODE REVIEW

File: moduleframework/helpfile_linter.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+from __future__ import print_function
+
+import os
+from moduleframework import common
+
+HELP_MD = "help.md"
+
+
+def get_help_md_file(dir_name):
+    helpmd_file = os.path.join(os.path.abspath(dir_name), "help", HELP_MD)
```
Comment (TomasTomecek) [inline moduleframework/helpfile_linter.py]: we should follow fedora guidelines here: https://fedoraproject.org/wiki/Container:Guidelines#Help_File

File: moduleframework/tools/dockerlint.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    def setUp(self):
+        # it is not intended just for docker, but just docker packages are
+        # actually properly signed
+        self.dp = dockerlinter.DockerfileLinter()
+        if self.dp.dockerfile is None:
+            self.skip()
+
+    def test_architecture_in_env_and_label_exists(self):
+        self.assertTrue(self.dp.get_docker_specific_env("ARCH="))
```
Comment (TomasTomecek) [inline moduleframework/tools/dockerlint.py]: it's not a requirement to have it as env var

File: moduleframework/tools/dockerlint.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+class DockerLint(container_avocado_test.ContainerAvocadoTest):
+    """
+    :avocado: enable
+    """
+
+    def testBasic(self):
+        self.start()
+        self.assertTrue("bin" in self.run("ls /").stdout)
+
+    def testContainerIsRunning(self):
```
Comment (TomasTomecek) [inline moduleframework/tools/dockerlint.py]: what's the purpose of this check?

File: moduleframework/tools/helpmd_lint.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        self.assertTrue(self.helpmd.get_tag("DESCRIPTION"))
+
+    def test_helpmd_usage(self):
+        self.assertTrue(self.helpmd.get_tag("USAGE"))
+
+    def test_helpmd_environment_variables(self):
+        self.assertTrue(self.helpmd.get_tag("ENVIRONMENT VARIABLES"))
+
+    def test_helpmd_security_implications(self):
+        self.assertTrue(self.helpmd.get_tag("SECURITY IMPLICATIONS"))
```
Comment (TomasTomecek) [inline moduleframework/tools/helpmd_lint.py]: how is this a requirement?

File: moduleframework/tools/helpmd_lint.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        self.assertTrue(self.helpmd.get_tag("NAME"))
+
+    def test_helpmd_description(self):
+        self.assertTrue(self.helpmd.get_tag("DESCRIPTION"))
+
+    def test_helpmd_usage(self):
+        self.assertTrue(self.helpmd.get_tag("USAGE"))
+
+    def test_helpmd_environment_variables(self):
+        self.assertTrue(self.helpmd.get_tag("ENVIRONMENT VARIABLES"))
```
Comment (TomasTomecek) [inline moduleframework/tools/helpmd_lint.py]: not all images operate with environment variables

File: requirements.txt [hunk 6]
```diff
@@ -5,3 +5,4 @@ dockerfile-parse
 modulemd
 netifaces
 pdc-client
+mistune
```
Comment (TomasTomecek) [inline requirements.txt]: this change seems to be unrelated to this PR

File: moduleframework/helpfile_linter.py (hunk 1 above)
Comment (phracek) [reply moduleframework/helpfile_linter.py]: Good to know, but this should be also documented in distgen. Nowadays, we are building `help.1` file into `./root/help.1` directory and help.md itself is on help/ directory.

File: moduleframework/tools/dockerlint.py (hunk 3 above)
Comment (phracek) [reply moduleframework/tools/dockerlint.py]: It was proposed at the beginning of MTF. I guess, just for sure, container is running.
@jscotka Is it correct? Can it be removed?

File: moduleframework/tools/helpmd_lint.py (hunk 4 above)
Comment (phracek) [reply moduleframework/tools/helpmd_lint.py]: I will modify it, so that if container expose some ports, then the ports  should be mentioned in this label.

File: moduleframework/tools/dockerlint.py (hunk 3 above)
Comment (jscotka) [reply moduleframework/tools/dockerlint.py]: sure, just example, it just test you are in container able to do something. nothing important.

Comment (TomasTomecek) [reply moduleframework/tools/dockerlint.py]: So can we remove it if it's just an example and not important?

File: moduleframework/dockerlinter.py [hunk 7]
```diff
@@ -161,11 +159,15 @@ def get_specific_label(self, label_name=None):
         if label_name is None:
             return []
         label_list = self.get_docker_labels()
-        return [label_name in label_list]
+        return [label_list[key] for key in label_list.keys() if label_name == key]
 
-    def check_baseruntime(self):
+    def check_from_is_first(self):
```
Comment (TomasTomecek) [inline moduleframework/dockerlinter.py]: Docs of this method and its implementation are two completely different things.

File: moduleframework/dockerlinter.py [hunk 8]
```diff
@@ -161,11 +159,15 @@ def get_specific_label(self, label_name=None):
...
         """
-        Function returns docker labels
-        :return: label dictionary
+        Function checks if FROM directive is really first directive.
+        Function ignores whitespaces and comments.
+        :return: True if FROM is first, False if FROM is not first directive
         """
-        return [x for x in self.docker_dict.get(FROM,[]) if "baseruntime/baseruntime" in x]
+        with open(self.dockerfile) as f:
```
Comment (TomasTomecek) [inline moduleframework/dockerlinter.py]: why parse it again when you use `DockerfileParser`?

File: moduleframework/helpfile_linter.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
+from __future__ import print_function
+
+import os
+from moduleframework import common
+
+HELP_MD = "help.md"
+
+
+def get_help_md_file(dockerfile):
```
Comment (TomasTomecek) [inline moduleframework/helpfile_linter.py]: still doesn't comply with Fedora Guidelines: https://fedoraproject.org/wiki/Container:Guidelines#Help_File

File: moduleframework/tools/dockerlint.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        if env:
+            label = self.dp.get_docker_specific_env(docker_env)
+        else:
+            label = self.dp.get_specific_label(docker_env)
+        if not label:
+            label_found = self.dp.get_specific_label(docker_label)
+        return label_found
+
+    def test_architecture_in_env_and_label_exists(self):
+        self.assertTrue(self._test_for_env_and_label("ARCH=", "architecture"))
```
Comment (TomasTomecek) [inline moduleframework/tools/dockerlint.py]: Can you please point me to guidelines which require dockerfiles to have environment variable `ARCH`?

File: moduleframework/tools/dockerlint.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        Function tests whether labels are set in modulemd YAML file properly.
+        :return:
+        """
+        llabels = self.getConfigModule().get('labels')
+        if llabels is None or len(llabels) == 0:
+            print("No labels defined in config to check")
+            self.cancel()
+        for key in self.getConfigModule()['labels']:
+            aaa = self.checkLabel(key, self.getConfigModule()['labels'][key])
+            print(">>>>>> ", aaa, key)
```
Comment (TomasTomecek) [inline moduleframework/tools/dockerlint.py]: probably a leftover from debugging

File: moduleframework/tools/dockerlint.py (hunk 10 above)
Comment (phracek) [reply moduleframework/tools/dockerlint.py]: see here: https://fedoraproject.org/wiki/Container:Guidelines#LABELS architecture is required.
But I got your point. We can only test if LABEL architecture is present.
My test was based on container-image-template: https://github.com/container-images/container-image-template/blob/master/Dockerfile.template#L5
There are mentioned as ARCH and architecture.
But I will fix it.

File: moduleframework/tools/dockerlint.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    #def test_from_correct_format(self):
+    #    self.assertTrue(self.dp.check_from_format())
+
+
+class DockerLint(container_avocado_test.ContainerAvocadoTest):
+    """
+    :avocado: enable
+    """
+
+    def testBasic(self):
```
[the remark this answers is not in the document]
Comment (phracek) [reply moduleframework/tools/dockerlint.py]: I guess, this is some kind of sanity check. But it can be removed.

File: moduleframework/dockerlinter.py (hunk 7 above)
Comment (phracek) [reply moduleframework/dockerlinter.py]: Fixed.

**Comment by phracek** [comment]
This Pull Request also adds checks like:
* FROM directive is valid
* USER has proper format, like only digits
* Dockerfile contains only one RUN instruction with dnf command.

File: moduleframework/tools/dockerlint.py (hunk 10 above)
Comment (TomasTomecek) [reply moduleframework/tools/dockerlint.py]: You are right, it's in there. I missed it, sorry.

Therefore we should check the presence of env vars and maybe produce a warning it they are missing.

File: moduleframework/dockerlinter.py [hunk 13]
```diff
@@ -161,11 +162,62 @@ def get_specific_label(self, label_name=None):
...
+        """
+        correct_format = False
+        struct = self.dfp_structure[0]
+        if struct.get(INSTRUCT) == 'FROM':
+            p = re.compile("^[a-z0-9.]+(\/[a-z0-9\D.]+)+$")
+            if p.search(struct.get('value')) is not None:
+                correct_format = True
+        return correct_format
+
+    def check_two_dnf_run_sections(self):
```
Comment (TomasTomecek) [inline moduleframework/dockerlinter.py]: This check does nothing with `dnf`. It should be about chained `RUN`s, so:

### bad
```
FROM fedora
RUN ls /
RUN cd /
```


### good
```
FROM fedora
RUN ls / && cd /
```

File: moduleframework/dockerlinter.py [hunk 14]
```diff
@@ -161,11 +162,62 @@ def get_specific_label(self, label_name=None):
...
+        value = 0
+        for struct in self.dfp_structure:
+            if struct.get(INSTRUCT) == RUN:
+                value += 1
+        if int(value) > 1:
+            return False
+        return True
+
+    def check_user_number(self):
```
Comment (TomasTomecek) [inline moduleframework/dockerlinter.py]: can be also string, see full spec: https://docs.docker.com/engine/reference/builder/#user

Comment (phracek) [reply moduleframework/dockerlinter.py]: Thanks I did not find it before implementation.
That means, container maintainer should know what he is writing in Dockerfile. We have no possibility how to check this.
I will remove it.

File: moduleframework/dockerlinter.py [hunk 15]
```diff
@@ -161,11 +166,72 @@ def get_specific_label(self, label_name=None):
...
+             FROM fedora
+             RUN ls / && cd /
+        :return: True if Dockerfile does not contain RUN instructions in rows
+                False if Dockerfile contains more RUN instructions in rows
+        """
+        value = 0
+        for struct in self.dfp_structure:
+            if struct.get(INSTRUCT) == RUN:
+                value += 1
+        if int(value) > 1:
```
Comment (TomasTomecek) [inline moduleframework/dockerlinter.py]: this seems to permit only a single `RUN` instruction, that's not good

Comment (phracek) [reply moduleframework/dockerlinter.py]: Well e.g.

~~~
FROM fedora
RUN dnf install foobar1 && dnf clean all
RUN ls /
RUN cd /
~~~
This should be just two RUN instructions.

~~~
FROM fedora
RUN dnf install foobar1
RUN dnf install clean all
RUN ls /
RUN cd /
~~~

This should be also two RUN instructions.

I would prefer again separate it into two tests. One for testing `dnf` and stuff around it and second one for testing rest instructions.

WDYT? I hate sort cuts.

Comment (TomasTomecek) [reply moduleframework/dockerlinter.py]: Sounds like a good idea since you correctly pointed out those are 2 separate use cases.

Bear in mind that current implementation just counts `RUN`s, so e.g. [this dockerfile](https://github.com/container-images/postfix/blob/master/Dockerfile) would not pass.

File: moduleframework/dockerlinter.py [hunk 16]
```diff
@@ -161,11 +167,95 @@ def get_specific_label(self, label_name=None):
...
+        """
+        correct_format = False
+        struct = self.dfp_structure[0]
+        if struct.get(INSTRUCT) == 'FROM':
+            p = re.compile("^[a-z0-9.]+(\/[a-z0-9\D.]+)+$")
+            if p.search(struct.get('value')) is not None:
+                correct_format = True
+        return correct_format
+
+    def check_chained_run_dnf_commands(self):
```
Comment (TomasTomecek) [inline moduleframework/dockerlinter.py]: This is a good test.

File: moduleframework/dockerlinter.py [hunk 17]
```diff
@@ -161,11 +167,95 @@ def get_specific_label(self, label_name=None):
...
+             RUN dnf clean all
+        GOOD example:
+             FROM fedora
+             RUN dnf install foobar1 && dnf clean all
+        :return: True if Dockerfile contains RUN dnf instructions in one row
+                False if Dockerfile contains RUN dnf instructions in more rows
+        """
+        value = 0
+        for struct in self.dfp_structure:
+            if struct.get(INSTRUCT) == RUN and "dnf" in struct.get("value"):
```
Comment (TomasTomecek) [inline moduleframework/dockerlinter.py]: what about yum?

File: moduleframework/dockerlinter.py [hunk 18]
```diff
@@ -161,11 +167,95 @@ def get_specific_label(self, label_name=None):
...
+                value += 1
+        if int(value) > 1:
+            return False
+        return True
+
+    def check_chained_run_rest_commands(self):
```
Comment (TomasTomecek) [inline moduleframework/dockerlinter.py]: I think this test may produce some false positives: sometimes it's convenient to have multiple RUN instructions (which operate on files added to the container)

File: moduleframework/helpfile_linter.py [hunk 19]
```diff
@@ -0,0 +1,49 @@
+from __future__ import print_function
+
+import os
+from moduleframework import common
+
+HELP_MD = "help.md"
+
+
+class HelpMDLinter(object):
```
[the remark this answers is not in the document]
Comment (phracek) [reply moduleframework/helpfile_linter.py]: Good for other PR.

[the remark this answers is not in the document]
Comment (phracek) [reply moduleframework/helpfile_linter.py]: https://github.com/fedora-modularity/meta-test-family/issues/138
